### PR TITLE
Persist individually deleted messages

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Carrying over pr #192 from @sschepens. 

Rebased on current master + commit 00fcab1e759de915a367b83e5089ed1c79ec3bd2

Added:
 * Maintain protobuf text format backward compatibility
 * If there are individually deleted message ranges to save, always store them in the ledger rather than the z-node (temp solution until binary format will be enabled)
 * Unit tests

### CODE REVIEW

**Comment by merlimat** [comment]
Ping @saandrews @rdhabalia

**Comment by rdhabalia** [comment]
yes, I will review it soon.

**Comment by merlimat** [comment]
No rush, just that I have the other changes for binary format in ML written on top of this :)

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpFindNewest.java [hunk 1]
```diff
@@ -15,13 +15,14 @@
...
-import org.apache.bookkeeper.mledger.AsyncCallbacks.ReadEntryCallback;
 import org.apache.bookkeeper.mledger.AsyncCallbacks.FindEntryCallback;
+import org.apache.bookkeeper.mledger.AsyncCallbacks.ReadEntryCallback;
 import org.apache.bookkeeper.mledger.Entry;
 import org.apache.bookkeeper.mledger.ManagedLedgerException;
 import org.apache.bookkeeper.mledger.Position;
 import org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.PositionBound;
-import com.google.common.base.Predicate;
+
+import java.util.function.Predicate;
```
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpFindNewest.java]: is there any specific reason to change this one?

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java [hunk 2]
```diff
@@ -2278,5 +2280,93 @@ public void testReopenMultipleTimes() throws Exception {
         c1 = ledger.openCursor("c1");
     }
 
+    @Test(timeOut = 20000)
```
Comment (rdhabalia) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java]: should we add a test-case which stores `individualDeletedPositions` into ledger and reads back as this PR does this functionality?

File: conf/broker.conf [hunk 3]
```diff
@@ -198,6 +198,8 @@ managedLedgerCursorMaxEntriesPerLedger=50000
 # Max time before triggering a rollover on a cursor ledger
 managedLedgerCursorRolloverTimeInSeconds=14400
 
+managedLedgerMaxUnackedRangesToPersist=1000
```
Comment (rdhabalia) [inline conf/broker.conf]: should we add the documentation?

File: pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java [hunk 4]
```diff
@@ -179,6 +179,11 @@
     private int managedLedgerCursorMaxEntriesPerLedger = 50000;
     // Max time before triggering a rollover on a cursor ledger
     private int managedLedgerCursorRolloverTimeInSeconds = 14400;
+    // Max number of entries to append to a ledger before triggering a rollover
+    // A ledger rollover is triggered on these conditions Either the max
+    // rollover time has been reached or max entries have been written to the
+    // ledged and at least min-time has passed
+    private int managedLedgerMaxUnackedRangesToPersist = 1000;
```
Comment (rdhabalia) [inline pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java]: I think we don't [persistPosition](https://github.com/merlimat/pulsar/blob/ecf450173d170a9324425c94f0480b8d3feb8c90/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L1914) only on `ledger rollover` but persist `markDeletePosition` based on `managedLedgerDefaultMarkDeleteRateLimit`

File: pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java [hunk 5]
```diff
@@ -179,6 +179,11 @@
     private int managedLedgerCursorMaxEntriesPerLedger = 50000;
     // Max time before triggering a rollover on a cursor ledger
     private int managedLedgerCursorRolloverTimeInSeconds = 14400;
+    // Max number of entries to append to a ledger before triggering a rollover
+    // A ledger rollover is triggered on these conditions Either the max
+    // rollover time has been reached or max entries have been written to the
```
Comment (rdhabalia) [inline pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java]: I think we don't [persistPosition](https://github.com/merlimat/pulsar/blob/ecf450173d170a9324425c94f0480b8d3feb8c90/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L1914) only on `ledger rollover` but persist `markDeletePosition` based on `managedLedgerDefaultMarkDeleteRateLimit`

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 6]
```diff
@@ -1825,39 +1883,67 @@ public void deleteComplete(int rc, Object ctx) {
...
+                        p = positionRange.upperEndpoint();
+                        nestedPositionBuilder.setLedgerId(p.getLedgerId());
+                        nestedPositionBuilder.setEntryId(p.getEntryId());
+                        messageRangeBuilder.setUpperEndpoint(nestedPositionBuilder.build());
+                        return messageRangeBuilder.build();
+                    })
+                    .collect(Collectors.toList());
+        } finally {
+            lock.readLock().unlock();
+        }
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: The ML protobuf classes are not generated with the custom protobuf, so they're not recyclable. I'd leave that for later, once we have completely phased out the text format.

Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: yes, that's correct.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java (hunk 2 above)
Comment (merlimat) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java]: There are 2 tests, one that closes the ledger/factory (simulating graceful shutdown) and  the other that directly opens a new ML factory, simulating a recovery after a crash.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpFindNewest.java (hunk 1 above)
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpFindNewest.java]: in general we should switch to JDK's own things when available, though in this case this is unrelated so I'll take it out

File: pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java (hunk 5 above)
Comment (merlimat) [inline pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/ServiceConfiguration.java]: Good point, this was copy-pasted from above config line. I'll fix it

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 7]
```diff
@@ -2025,9 +2111,9 @@ void asyncDeleteCursorLedger() {
      */
     private static boolean isBkErrorNotRecoverable(int rc) {
         switch (rc) {
-        case Code.NoSuchLedgerExistsException:
-        case Code.ReadException:
-        case Code.LedgerRecoveryException:
+        case BKException.Code.NoSuchLedgerExistsException:
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: No reason in this PR, will fix it

File: conf/broker.conf (hunk 3 above)
Comment (merlimat) [inline conf/broker.conf]: Yes, we should

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 7 above)
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: Actually, it's good to disambiguate between ZK & BK `Code` classes

**Comment by merlimat** [comment]
@rdhabalia Updated

**Pull request merged**